LLAP/llap.sh

```diff
@@ -0,0 +1,478 @@
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+set -euxo pipefail
+
+##install xml modifiction tool....
+apt-get install -y xmlstarlet
+
+readonly NOT_SUPPORTED_MESSAGE="LLAP initialization action is not supported on Dataproc ${DATAPROC_VERSION}."
+[[ $DATAPROC_VERSION != 2.* ]] && echo "$NOT_SUPPORTED_MESSAGE" && exit 1
+
+# Variables for running this script
+readonly ROLE="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
+readonly LLAP_MASTER_FQDN="$(/usr/share/google/get_metadata_value attributes/dataproc-master)"
+readonly WORKER_NODE_COUNT=$(/usr/share/google/get_metadata_value attributes/dataproc-worker-count)
+readonly NODE_MANAGER_vCPU=$(bdconfig get_property_value --configuration_file='/etc/hadoop/conf/yarn-site.xml' --name yarn.nodemanager.resource.cpu-vcores)
+readonly NODE_MANAGER_MEMORY=$(bdconfig get_property_value --configuration_file='/etc/hadoop/conf/yarn-site.xml' --name yarn.nodemanager.resource.memory-mb)
+readonly YARN_MAX_CONTAINER_MEMORY=$(bdconfig get_property_value --configuration_file='/etc/hadoop/conf/yarn-site.xml' --name yarn.scheduler.maximum-allocation-mb)
+readonly ADDITIONAL_MASTER=$(/usr/share/google/get_metadata_value attributes/dataproc-master-additional)
+readonly HAS_SSD=$(/usr/share/google/get_metadata_value attributes/ssd)
+readonly NUM_LLAP_NODES=$(/usr/share/google/get_metadata_value attributes/num-llap-nodes)
+readonly HADOOP_CONF_DIR='/etc/hadoop/conf'
+readonly HIVE_CONF_DIR='/etc/hive/conf'
+
+
+function pre_flight_checks(){
+
+##check for bad configurations
+if [[ "${NUM_LLAP_NODES}" -ge "${WORKER_NODE_COUNT}" ]]; then
+    echo "LLAP node count equals total worker count. There are no nodes to support Tez AM's. Please reduce LLAP instance count and re-deploy." && exit 1
+fi
+
+###check to see if HA or not
+if [[ -n "$ADDITIONAL_MASTER" ]]; then
+    IS_HA=true
+else
+    IS_HA=false
+fi
+}
+
+##add xml doc tool for editing hadoop configuration files
+function configure_yarn_site(){
+    echo "configure yarn-site.xml..."
+
+    xmlstarlet edit --inplace --omit-decl \
+    --update '//configuration/property[name="yarn.application.classpath"]/value' \
+    -x 'concat(.,",\$HADOOP_CONF_DIR,/usr/local/share/google/dataproc/lib/*,/usr/lib/hadoop/*,/usr/lib/hadoop-mapreduce/*,/usr/lib/hadoop/lib/*,/usr/lib/hadoop-hdfs/*,/usr/lib/hadoop-hdfs/lib/*,/usr/lib/hadoop-yarn/*,/usr/lib/hadoop-yarn/lib/*,/usr/lib/tez/*,/usr/lib/tez/lib/*")' \
+    /etc/hadoop/conf/yarn-site.xml
+
+    if [[ "${NODE_MANAGER_MEMORY}" != "${YARN_MAX_CONTAINER_MEMORY}" ]]; then
+	   echo "not configured properly..."
+    fi
+}
+
+###add configurations to hive-site for LLAP
+function configure_hive_site(){
+
+    ##different configuration if HA
+    echo "configure hive-site.xml...."
+    if $IS_HA; then
```

Comment:
Can replace this with `if [[ -n "$ADDITIONAL_MASTER" ]]; then`. No need for separate variable for this.

Reply:
fixed